File: Source/Tools/Editor/SceneTab.cpp
//
// Copyright (c) 2008-2017 the Urho3D project.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

#include "SceneTab.h"
#include "EditorEvents.h"
#include "SceneSettings.h"
#include <Toolbox/Scene/DebugCameraController.h>
#include <ImGui/imgui_internal.h>
#include <ImGuizmo/ImGuizmo.h>
#include <IconFontCppHeaders/IconsFontAwesome.h>


namespace Urho3D
{

SceneTab::SceneTab(Context* context, StringHash id, const String& afterDockName, ui::DockSlot_ position)
    : SceneView(context, {0, 0, 1024, 768})
    , gizmo_(context)
    , inspector_(context)
    , placeAfter_(afterDockName)
    , placePosition_(position)
    , id_(id)
{
    SetTitle(title_);

    settings_ = new SceneSettings(context);
    effectSettings_ = new SceneEffects(this);

    SubscribeToEvent(this, E_EDITORSELECTIONCHANGED, std::bind(&SceneTab::OnNodeSelectionChanged, this));
    SubscribeToEvent(effectSettings_, E_EDITORSCENEEFFECTSCHANGED, std::bind(&AttributeInspector::CopyEffectsFrom,
                                                                             &inspector_, viewport_));
}

SceneTab::~SceneTab() = default;

void SceneTab::SetSize(const IntRect& rect)
{
    SceneView::SetSize(rect);
    gizmo_.SetScreenRect(rect);
}

bool SceneTab::RenderWindow()
{
    bool open = true;
    auto& style = ui::GetStyle();

    if (GetInput()->IsMouseVisible())
        lastMousePosition_ = GetInput()->GetMousePosition();

    ui::SetNextDockPos(placeAfter_.CString(), placePosition_, ImGuiCond_FirstUseEver);
    if (ui::BeginDock(uniqueTitle_.CString(), &open, windowFlags_))
    {
        // Focus window when appearing
        if (!isRendered_)
        {
            ui::SetWindowFocus();
            isRendered_ = true;
            effectSettings_->Prepare(true);
        }

        ImGuizmo::SetDrawlist();
        ui::SetCursorPos(ui::GetCursorPos() - style.WindowPadding);
        ui::Image(texture_, ToImGui(rect_.Size()));

        if (rect_.IsInside(lastMousePosition_) == INSIDE)
        {
            if (!ui::IsWindowFocused() && ui::IsItemHovered() && GetInput()->GetMouseButtonDown(MOUSEB_RIGHT))
                ui::SetWindowFocus();

            if (ui::IsDockActive())
                isActive_ = ui::IsWindowFocused();
            else
                isActive_ = false;
        }
        else
            isActive_ = false;

        camera_->GetComponent<DebugCameraController>()->SetEnabled(isActive_);

        gizmo_.ManipulateSelection(GetCamera());

        // Update scene view rect according to window position
        // if (!GetInput()->GetMouseButtonDown(MOUSEB_LEFT))
        {
            auto titlebarHeight = ui::GetCurrentContext()->CurrentWindow->TitleBarHeight();
            auto pos = ui::GetWindowPos();
            pos.y += titlebarHeight;
            auto size = ui::GetWindowSize();
            size.y -= titlebarHeight;
            if (size.x > 0 && size.y > 0)
            {
                IntRect newRect(ToIntVector2(pos), ToIntVector2(pos + size));
                SetSize(newRect);
            }
        }

        if (ui::IsItemHovered())
        {
            // Prevent dragging window when scene view is clicked.
            windowFlags_ = ImGuiWindowFlags_NoMove;

            // Handle object selection.
            if (!gizmo_.IsActive() && GetInput()->GetMouseButtonPress(MOUSEB_LEFT))
            {
                IntVector2 pos = GetInput()->GetMousePosition();
                pos -= rect_.Min();

                Ray cameraRay = GetCamera()->GetScreenRay((float)pos.x_ / rect_.Width(), (float)pos.y_ / rect_.Height());
                // Pick only geometry objects, not eg. zones or lights, only get the first (closest) hit
                PODVector<RayQueryResult> results;

                RayOctreeQuery query(results, cameraRay, RAY_TRIANGLE, M_INFINITY, DRAWABLE_GEOMETRY);
                scene_->GetComponent<Octree>()->RaycastSingle(query);

                if (!results.Size())
                {
                    // When object geometry was not hit by a ray - query for object bounding box.
                    RayOctreeQuery query2(results, cameraRay, RAY_OBB, M_INFINITY, DRAWABLE_GEOMETRY);
                    scene_->GetComponent<Octree>()->RaycastSingle(query2);
                }

                if (results.Size())
                {
                    WeakPtr<Node> clickNode(results[0].drawable_->GetNode());
                    if (!GetInput()->GetKeyDown(KEY_CTRL))
                        UnselectAll();

                    ToggleSelection(clickNode);
                }
                else
                    UnselectAll();
            }
        }
        else
            windowFlags_ = 0;

        const auto tabContextMenuTitle = "SceneTab context menu";
        if (ui::IsDockTabHovered() && GetInput()->GetMouseButtonPress(MOUSEB_RIGHT))
            ui::OpenPopup(tabContextMenuTitle);
        if (ui::BeginPopup(tabContextMenuTitle))
        {
            if (ui::MenuItem("Save"))
                SaveScene();

            ui::Separator();

            if (ui::MenuItem("Close"))
                open = false;

            ui::EndPopup();
        }
    }
    else
    {
        isActive_ = false;
        isRendered_ = false;
    }
    ui::EndDock();

    return open;
}

void SceneTab::LoadScene(const String& filePath)
{
    if (filePath.Empty())
        return;

    if (filePath.EndsWith(".xml", false))
    {
        if (scene_->LoadXML(GetCache()->GetResource<XMLFile>(filePath)->GetRoot()))
        {
            path_ = filePath;
            CreateObjects();
        }
        else
            URHO3D_LOGERRORF("Loading scene %s failed", GetFileName(filePath).CString());
    }
    else if (filePath.EndsWith(".json", false))
    {
        if (scene_->LoadJSON(GetCache()->GetResource<JSONFile>(filePath)->GetRoot()))
        {
            path_ = filePath;
            CreateObjects();
        }
        else
            URHO3D_LOGERRORF("Loading scene %s failed", GetFileName(filePath).CString());
    }
    else
        URHO3D_LOGERRORF("Unknown scene file format %s", GetExtension(filePath).CString());
}

bool SceneTab::SaveScene(const String& filePath)
{
    auto resourcePath = filePath.Empty() ? path_ : filePath;
    auto fullPath = GetCache()->GetResourceFileName(resourcePath);
    File file(context_, fullPath, FILE_WRITE);
    bool result = false;

    float elapsed = 0;
    if (!settings_->saveElapsedTime_)
    {
        elapsed = scene_->GetElapsedTime();
        scene_->SetElapsedTime(0);
    }

    if (fullPath.EndsWith(".xml", false))
        result = scene_->SaveXML(file);
    else if (fullPath.EndsWith(".json", false))
        result = scene_->SaveJSON(file);

    if (!settings_->saveElapsedTime_)
        scene_->SetElapsedTime(elapsed);

    if (result)
    {
        if (!filePath.Empty())
            path_ = filePath;
    }
    else
        URHO3D_LOGERRORF("Saving scene to %s failed.", resourcePath.CString());

    return result;
}

void SceneTab::CreateObjects()
{
    SceneView::CreateObjects();
    camera_->CreateComponent<DebugCameraController>();
}

void SceneTab::Select(Node* node)
{
    if (gizmo_.Select(node))
    {
        using namespace EditorSelectionChanged;
        SendEvent(E_EDITORSELECTIONCHANGED, P_SCENETAB, this);
    }
}

void SceneTab::Unselect(Node* node)
{
    if (gizmo_.Unselect(node))
    {
        using namespace EditorSelectionChanged;
        SendEvent(E_EDITORSELECTIONCHANGED, P_SCENETAB, this);
    }
}

void SceneTab::ToggleSelection(Node* node)
{
    gizmo_.ToggleSelection(node);
    using namespace EditorSelectionChanged;
    SendEvent(E_EDITORSELECTIONCHANGED, P_SCENETAB, this);
}

void SceneTab::UnselectAll()
{
    if (gizmo_.UnselectAll())
    {
        using namespace EditorSelectionChanged;
        SendEvent(E_EDITORSELECTIONCHANGED, P_SCENETAB, this);
    }
}

const Vector<WeakPtr<Node>>& SceneTab::GetSelection() const
{
    return gizmo_.GetSelection();
}

void SceneTab::RenderGizmoButtons()
{
    const auto& style = ui::GetStyle();

    auto drawGizmoOperationButton = [&](GizmoOperation operation, const char* icon, const char* tooltip)
    {
        if (gizmo_.GetOperation() == operation)
            ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_ButtonActive]);
        else
            ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_Button]);
        if (ui::ButtonEx(icon, {20, 20}, ImGuiButtonFlags_PressedOnClick))
            gizmo_.SetOperation(operation);
        ui::PopStyleColor();
        ui::SameLine();
        if (ui::IsItemHovered())
            ui::SetTooltip("%s", tooltip);
    };

    auto drawGizmoTransformButton = [&](TransformSpace transformSpace, const char* icon, const char* tooltip)
    {
        if (gizmo_.GetTransformSpace() == transformSpace)
            ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_ButtonActive]);
        else
            ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_Button]);
        if (ui::ButtonEx(icon, {20, 20}, ImGuiButtonFlags_PressedOnClick))
            gizmo_.SetTransformSpace(transformSpace);
        ui::PopStyleColor();
        ui::SameLine();
        if (ui::IsItemHovered())
            ui::SetTooltip("%s", tooltip);
    };

    drawGizmoOperationButton(GIZMOOP_TRANSLATE, ICON_FA_ARROWS, "Translate");
    drawGizmoOperationButton(GIZMOOP_ROTATE, ICON_FA_REPEAT, "Rotate");
    drawGizmoOperationButton(GIZMOOP_SCALE, ICON_FA_ARROWS_ALT, "Scale");
    ui::TextUnformatted("|");
    ui::SameLine();
    drawGizmoTransformButton(TS_WORLD, ICON_FA_ARROWS, "World");
    drawGizmoTransformButton(TS_LOCAL, ICON_FA_ARROWS_ALT, "Local");
    ui::TextUnformatted("|");
    ui::SameLine();


    auto light = camera_->GetComponent<Light>();
    if (light->IsEnabled())
        ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_ButtonActive]);
    else
        ui::PushStyleColor(ImGuiCol_Button, style.Colors[ImGuiCol_Button]);
    if (ui::Button(ICON_FA_LIGHTBULB_O, {20, 20}))
        light->SetEnabled(!light->IsEnabled());
    ui::PopStyleColor();
    ui::SameLine();
    if (ui::IsItemHovered())
        ui::SetTooltip("Camera Headlight");
}

bool SceneTab::IsSelected(Node* node) const
{
    return gizmo_.IsSelected(node);
}

void SceneTab::OnNodeSelectionChanged()
{
    using namespace EditorSelectionChanged;
    const auto& selection = GetSelection();
    if (selection.Size() == 1)
    {
        const auto& node = selection.Front();
        const auto& components = node->GetComponents();
        if (!components.Empty())
            selectedComponent_ = components.Front();
        else
            selectedComponent_ = nullptr;
    }
    else
        selectedComponent_ = nullptr;
}

void SceneTab::RenderInspector()
{
    // TODO: inspector for multi-selection.
    if (GetSelection().Size() == 1)
    {
        auto node = GetSelection().Front();
        PODVector<Serializable*> items;
        items.Push(dynamic_cast<Serializable*>(node.Get()));
        if (node == scene_)
        {
            effectSettings_->Prepare();
            items.Push(settings_.Get());
            items.Push(effectSettings_.Get());
        }
        if (!selectedComponent_.Expired())
            items.Push(dynamic_cast<Serializable*>(selectedComponent_.Get()));
        inspector_.RenderAttributes(items);
    }
}

void SceneTab::RenderSceneNodeTree(Node* node)
{
    ImGuiTreeNodeFlags flags = ImGuiTreeNodeFlags_OpenOnArrow;
    if (node == nullptr)
    {
        flags |= ImGuiTreeNodeFlags_DefaultOpen;
        node = scene_;
    }

    if (node->IsTemporary())
        return;

    String name = ToString("%s (%d)", (node->GetName().Empty() ? node->GetTypeName() : node->GetName()).CString(), node->GetID());
    bool isSelected = IsSelected(node);

    if (isSelected)
        flags |= ImGuiTreeNodeFlags_Selected;

    auto opened = ui::TreeNodeEx(name.CString(), flags);

    if (ui::IsItemClicked(0))
    {
        if (!GetInput()->GetKeyDown(KEY_CTRL))
            UnselectAll();
        ToggleSelection(node);
    }

    if (opened)
    {
        for (auto& component: node->GetComponents())
        {
            if (component->IsTemporary())
                continue;

            bool selected = selectedComponent_ == component;
            if (ui::Selectable(component->GetTypeName().CString(), selected))
            {
                UnselectAll();
                ToggleSelection(node);
                selectedComponent_ = component;
            }
        }

        for (auto& child: node->GetChildren())
            RenderSceneNodeTree(child);
        ui::TreePop();
    }
}

void SceneTab::LoadProject(XMLElement scene)
{
    id_ = StringHash(ToUInt(scene.GetAttribute("id"), 16));
    SetTitle(scene.GetAttribute("title"));
    LoadScene(scene.GetAttribute("path"));

    auto camera = scene.GetChild("camera");
    if (camera.NotNull())
    {
        if (auto position = camera.GetChild("position"))
            camera_->SetPosition(position.GetVariant().GetVector3());
        if (auto rotation = camera.GetChild("rotation"))
            camera_->SetRotation(rotation.GetVariant().GetQuaternion());
        if (auto light = camera.GetChild("light"))
            camera_->GetComponent<Light>()->SetEnabled(light.GetVariant().GetBool());
    }

    settings_->LoadProject(scene);
    effectSettings_->LoadProject(scene);
}

void SceneTab::SaveProject(XMLElement scene) const
{
    scene.SetAttribute("id", id_.ToString().CString());
    scene.SetAttribute("title", title_);
    scene.SetAttribute("path", path_);

    auto camera = scene.CreateChild("camera");
    camera.CreateChild("position").SetVariant(camera_->GetPosition());
    camera.CreateChild("rotation").SetVariant(camera_->GetRotation());
    camera.CreateChild("light").SetVariant(camera_->GetComponent<Light>()->IsEnabled());

    settings_->SaveProject(scene);
    effectSettings_->SaveProject(scene);
}

void SceneTab::SetTitle(const String& title)
{
    title_ = title;
    uniqueTitle_ = ToString("%s###%s", title.CString(), id_.ToString().CString());
}

void SceneTab::ClearCachedPaths()
{
    path_.Clear();
}

}

File: Source/Tools/Editor.Net/Tabs/SceneTab.cs
//
// Copyright (c) 2018 Rokas Kupstys
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

using System;
using System.Collections.Generic;
using System.Linq;
using Editor.Events;
using IconFonts;
using Urho3D;
using ImGui;
using Urho3D.Events;


namespace Editor.Tabs
{
    public class SceneTab : Tab, IInspectable, IHierarchyProvider
    {
        private readonly SceneView _view;
        private readonly Gizmo _gizmo;
        private readonly VariantMap _eventArgs = new VariantMap();
        private bool _wasFocused;
        private bool _isMouseHoveringViewport;
        private Component _selectedComponent;
        private readonly IconCache _iconCache;

        public SceneTab(Context context, string title, Vector2? initialSize = null, string placeNextToDock = null,
            DockSlot slot = DockSlot.SlotNone) : base(context, title, initialSize, placeNextToDock, slot)
        {
            _iconCache = GetSubsystem<IconCache>();
            WindowFlags = WindowFlags.NoScrollbar;
            _view = new SceneView(Context);
            _gizmo = new Gizmo(Context);
            _view.Scene.LoadXml(Cache.GetResource<XMLFile>("Scenes/SceneLoadExample.xml").GetRoot());
            CreateObjects();

            SubscribeToEvent<Update>(OnUpdate);
            SubscribeToEvent<PostUpdate>(args => RenderNodeContextMenu());
            SubscribeToEvent<GizmoSelectionChanged>(args => { _selectedComponent = null; });

            _eventArgs.Clear();
            _eventArgs[InspectHierarchy.HierarchyProvider] = Variant.FromObject(this);
            SendEvent<InspectHierarchy>(_eventArgs);
        }

        private void RenderToolbar()
        {
            ui.PushStyleVar(StyleVar.FrameRounding, 0);

            if (Widgets.EditorToolbarButton(FontAwesome.Undo, "Undo"))
            {
            }

            if (Widgets.EditorToolbarButton(FontAwesome.Repeat, "Redo"))
            {
            }

            ui.SameLine(0, 3);

            if (Widgets.EditorToolbarButton(FontAwesome.Arrows, "Translate", _gizmo.Operation == GizmoOperation.Translate))
                _gizmo.Operation = GizmoOperation.Translate;
            if (Widgets.EditorToolbarButton(FontAwesome.Repeat, "Rotate", _gizmo.Operation == GizmoOperation.Rotate))
                _gizmo.Operation = GizmoOperation.Rotate;
            if (Widgets.EditorToolbarButton(FontAwesome.ArrowsAlt, "Scale", _gizmo.Operation == GizmoOperation.Scale))
                _gizmo.Operation = GizmoOperation.Scale;

            ui.SameLine(0, 3);

            if (Widgets.EditorToolbarButton(FontAwesome.ArrowsAlt, "World", _gizmo.TransformSpace == TransformSpace.World))
                _gizmo.TransformSpace = TransformSpace.World;
            if (Widgets.EditorToolbarButton(FontAwesome.ArrowsAlt, "Local", _gizmo.TransformSpace == TransformSpace.Local))
                _gizmo.TransformSpace = TransformSpace.Local;

            ui.SameLine(0, 3);

            var light = _view.Camera.Node.GetComponent<Light>();
            if (light != null)
            {
                if (Widgets.EditorToolbarButton(FontAwesome.LightbulbO, "Camera Headlight", light.IsEnabled))
                    light.IsEnabled = !light.IsEnabled;
            }

            ui.SameLine(0, 3);

            ui.PopStyleVar();
            ui.NewLine();
        }

        protected override void Render()
        {
            RenderToolbar();

            var padding = ui.GetStyle().FramePadding;
            var spacing = ui.GetStyle().ItemSpacing;
            var size = new System.Numerics.Vector2(ui.GetWindowWidth() - padding.X - spacing.X,
                ui.GetWindowHeight() - ui.GetCursorPos().Y - spacing.Y);
            if (ui.BeginChild(0xF00, size, false, WindowFlags.Default))
            {
                var viewSize = ui.GetWindowSize();
                _view.Size = viewSize;
                var viewPos = ui.GetCursorScreenPos();
                ui.Image(_view.Texture.NativeInstance, viewSize, Vector2.Zero, Vector2.One, Vector4.One, Vector4.Zero);
                if (Input.IsMouseVisible)
                    _isMouseHoveringViewport = ui.IsItemHovered(HoveredFlags.Default);

                _gizmo.SetScreenRect(viewPos, viewSize);

                var isClickedLeft = Input.GetMouseButtonClick((int) MouseButton.Left) && ui.IsItemHovered(HoveredFlags.AllowWhenBlockedByPopup);
                var isClickedRight = Input.GetMouseButtonClick((int) MouseButton.Right) && ui.IsItemHovered(HoveredFlags.AllowWhenBlockedByPopup);

                _gizmo.ManipulateSelection(_view.Camera);

                if (ui.IsWindowHovered())
                    WindowFlags |= WindowFlags.NoMove;
                else
                    WindowFlags &= ~WindowFlags.NoMove;

                if (!_gizmo.IsActive && (isClickedLeft || isClickedRight) && Input.IsMouseVisible)
                {
                    // Handle object selection.
                    var pos = Input.MousePosition - new IntVector2(viewPos);

                    var cameraRay = _view.Camera.GetScreenRay(pos.X / size.X, pos.Y / size.Y);
                    // Pick only geometry objects, not eg. zones or lights, only get the first (closest) hit

                    var query = new RayOctreeQuery(cameraRay, RayQueryLevel.Triangle, float.PositiveInfinity, DrawableFlags.Geometry);
                    _view.Scene.GetComponent<Octree>().RaycastSingle(query);

                    if (query.Result.Length == 0)
                    {
                        // When object geometry was not hit by a ray - query for object bounding box.
                        query = new RayOctreeQuery(cameraRay, RayQueryLevel.Obb, float.PositiveInfinity, DrawableFlags.Geometry);
                        _view.Scene.GetComponent<Octree>().RaycastSingle(query);
                    }

                    if (query.Result.Length > 0)
                    {
                        var clickNode = query.Result[0].Drawable.Node;
                        // Temporary nodes can not be selected.
                        while (clickNode != null && clickNode.HasTag("__EDITOR_OBJECT__"))
                            clickNode = clickNode.Parent;

                        if (clickNode != null)
                        {
                            var appendSelection = Input.GetKeyDown(InputEvents.KeyCtrl);
                            if (!appendSelection)
                                _gizmo.UnselectAll();
                            _gizmo.ToggleSelection(clickNode);

                            // Notify inspector
                            _eventArgs.Clear();
                            _eventArgs[InspectItem.Inspectable] = Variant.FromObject(this);
                            SendEvent<InspectItem>(_eventArgs);

                            if (isClickedRight)
                                ui.OpenPopup("Node context menu");
                        }
                    }
                    else
                        _gizmo.UnselectAll();
                }

                RenderNodeContextMenu();
            }
            ui.EndChild();
        }

        private void CreateObjects()
        {
            _view.CreateObjects();
            _view.Camera.Node.GetOrCreateComponent<DebugCameraController>();
        }

        private void OnUpdate(VariantMap args)
        {
            if (IsDockActive && _isMouseHoveringViewport)
            {
                _view.Camera.GetComponent<DebugCameraController>().Update(args[Update.TimeStep].Float);
                if (!_wasFocused)
                {
                    _wasFocused = true;
                    _eventArgs.Clear();
                    _eventArgs[InspectHierarchy.HierarchyProvider] = Variant.FromObject(this);
                    SendEvent<InspectHierarchy>(_eventArgs);
                }
            }
            else
                _wasFocused = false;
        }

        public Serializable[] GetInspectableObjects()
        {
            if (_gizmo.Selection == null)
                return new Serializable[0];

            var inspectables = _gizmo.Selection.Cast<Serializable>().ToList();
            if (_selectedComponent != null)
                inspectables.Add(_selectedComponent);

            return inspectables.ToArray();
        }

        private void RenderNodeTree(Node node)
        {
            if (node.IsTemporary)
                return;

            var flags = TreeNodeFlags.OpenOnArrow;
            if (node.Parent == null)
                flags |= TreeNodeFlags.DefaultOpen;

            var name = (node.Name.Length == 0 ? node.GetType().Name : node.Name) + $" ({node.Id})";
            var isSelected = _gizmo.IsSelected(node) || _selectedComponent != null && _selectedComponent.Node == node;

            if (isSelected)
                flags |= TreeNodeFlags.Selected;

            _iconCache.RenderIcon("Node");
            ui.SameLine();

            var opened = ui.TreeNodeEx(name, flags);
            if (!opened)
            {
                // If TreeNode above is opened, it pushes it's label as an ID to the stack. However if it is not open then no
                // ID is pushed. This creates a situation where context menu is not properly attached to said tree node due to
                // missing ID on the stack. To correct this we ensure that ID is always pushed. This allows us to show context
                // menus even for closed tree nodes.
                ui.PushID(name);
            }

            if (ui.IsItemHovered(HoveredFlags.AllowWhenBlockedByPopup))
            {
                if (ImGui.SystemUi.IsMouseClicked(MouseButton.Left))
                {
                    if (!Input.GetKeyDown(InputEvents.KeyCtrl))
                        _gizmo.UnselectAll();
                    _gizmo.ToggleSelection(node);
                }
                else if (ImGui.SystemUi.IsMouseClicked(MouseButton.Right))
                {
                    _gizmo.UnselectAll();
                    _gizmo.ToggleSelection(node);
                    ui.OpenPopup("Node context menu");
                }
            }

            RenderNodeContextMenu();

            if (opened)
            {
                foreach (var component in node.Components)
                {
                    if (component.IsTemporary)
                        continue;

                    ui.PushID(component.NativeInstance.ToInt32());

                    _iconCache.RenderIcon(component.GetType().Name);
                    ui.SameLine();

                    var selected = _selectedComponent != null && _selectedComponent == component;
                    selected = ui.Selectable(component.GetType().Name, selected);

                    if (ImGui.SystemUi.IsMouseClicked(MouseButton.Right) && ui.IsItemHovered(HoveredFlags.AllowWhenBlockedByPopup))
                    {
                        selected = true;
                        ui.OpenPopup("Component context menu");
                    }

                    if (selected)
                    {
                        _gizmo.UnselectAll();
                        _gizmo.ToggleSelection(node);
                        _selectedComponent = component;
                    }

                    if (ui.BeginPopup("Component context menu"))
                    {
                        if (ui.MenuItem("Remove"))
                            component.Remove();
                        ui.EndPopup();
                    }

                    ui.PopID();
                }

                // Do not use element->GetChildren() because child may be deleted during this loop.
                foreach (var child in node.Children)
                    RenderNodeTree(child);

                ui.TreePop();
            }
            else
                ui.PopID();
        }

        private void RenderNodeContextMenu()
        {
            if (ui.BeginPopup("Node context menu") /*&& !scenePlaying_*/)
            {
                if (Input.GetKeyPress(InputEvents.KeyEscape) || !Input.IsMouseVisible)
                {
                    // Close when interacting with scene camera.
                    ui.CloseCurrentPopup();
                    ui.EndPopup();
                    return;
                }

                var isAlternative = Input.GetKeyDown(InputEvents.KeyShift);

                if (ui.MenuItem(isAlternative ? "Create Child (Local)" : "Create Child"))
                {
                    foreach (var selectedNode in _gizmo.Selection)
                    {
                        var newNode = selectedNode.CreateChild(null,
                            isAlternative ? CreateMode.Local : CreateMode.Replicated);
                        _gizmo.Select(newNode);
                        _selectedComponent = null;
                    }
                }

                if (ui.BeginMenu(isAlternative ? "Create Component (Local)" : "Create Component"))
                {
                    var categories = Context.GetObjectCategories();

                    foreach (var category in categories)
                    {
                        if (category == "UI")
                            // Scene nodes can not contain UI elements
                            continue;

                        if (ui.BeginMenu(category))
                        {
                            var components = Context.GetObjectsByCategory(category);
                            Array.Sort(components, StringComparer.InvariantCulture);

                            foreach (var component in components)
                            {
                                _iconCache.RenderIcon(component);
                                ui.SameLine();

                                if (ui.MenuItem(component))
                                {
                                    foreach (var node in _gizmo.Selection)
                                    {
                                        node.CreateComponent(new StringHash(component),
                                            isAlternative ? CreateMode.Local : CreateMode.Replicated);
                                    }
                                }
                            }
                            ui.EndMenu();
                        }
                    }
                    ui.EndMenu();
                }

                ui.Separator();

                if (ui.MenuItem("Remove"))
                {
                    _selectedComponent?.Remove();
                    _selectedComponent = null;
                    foreach (var selected in _gizmo.Selection)
                        selected.Remove();
                    _gizmo.UnselectAll();
                }

                ui.EndPopup();
            }
        }

        public void RenderHierarchy()
        {
            ui.PushStyleVar(StyleVar.IndentSpacing, 10);
            try
            {
                RenderNodeTree(_view.Scene);
            }
            finally
            {
                ui.PopStyleVar();
            }
        }
    }
}
